icaruscode/TPC/Utilities/DetectorPropertiesICARUSClockOffsetMC.cxx

```diff
@@ -0,0 +1,483 @@
+  double DetectorPropertiesICARUSClockOffsetMC::BirksCorrection(double dQdx) const
+  {
+    return BirksCorrection(dQdx, Efield());
+  }
+  double DetectorPropertiesICARUSClockOffsetMC::BirksCorrection(double dQdx, double E_field) const
+  {
+    // Correction for charge quenching using parameterization from
+    // S.Amoruso et al., NIM A 523 (2004) 275
+
+    constexpr double A3t = util::kRecombA;
+    double K3t = util::kRecombk;                                    // in KV/cm*(g/cm^2)/MeV
+    double const rho = Density();                                   // LAr density in g/cm^3
+    constexpr double Wion = 1000. / util::kGeVToElectrons;          // 23.6 eV = 1e, Wion in MeV/e
+    K3t /= rho;                                                     // KV/MeV
+    double const dEdx = dQdx / (A3t / Wion - K3t / E_field * dQdx); // MeV/cm
+
+    return dEdx;
+  }
+
+  //----------------------------------------------------------------------------------
+  // Modified Box model correction
+  double DetectorPropertiesICARUSClockOffsetMC::ModBoxCorrection(double dQdx) const
+  {
+    return ModBoxCorrection(dQdx, Efield());
+  }
+  double DetectorPropertiesICARUSClockOffsetMC::ModBoxCorrection(double dQdx, double E_field) const
+  {
+    // Modified Box model correction has better behavior than the Birks
+    // correction at high values of dQ/dx.
+    double const rho = Density();                          // LAr density in g/cm^3
+    constexpr double Wion = 1000. / util::kGeVToElectrons; // 23.6 eV = 1e, Wion in MeV/e
+    double const Beta = fModBoxB / (rho * E_field);
+    double const Alpha = fModBoxA;
+    double const dEdx = (exp(Beta * Wion * dQdx) - Alpha) / Beta;
+
+    return dEdx;
+  }
+
+  //--------------------------------------------------------------------
+  //  x<--> ticks conversion methods
+  //
+  //  Ben Jones April 2012,
+  //  based on code by Herb Greenlee in SpacePointService
+  //
+
+  //--------------------------------------------------------------------
+  // Recalculate x<-->ticks conversion parameters from detector constants
+
+  DetectorPropertiesData DetectorPropertiesICARUSClockOffsetMC::DataFor(
+    detinfo::DetectorClocksData const& clock_data) const
+  {
+    double const samplingRate = sampling_rate(clock_data);
+    double const efield = Efield();
+    double const temperature = Temperature();
+    double const driftVelocity = DriftVelocity(efield, temperature);
+    double const x_ticks_coefficient = 0.001 * driftVelocity * samplingRate;
+
+    double extraOffset = 0.;
+    if ( clock_data.TriggerTime() > std::numeric_limits<double>::lowest() + std::numeric_limits<double>::epsilon() &&
+         clock_data.TriggerTime() < std::numeric_limits<double>::max() - std::numeric_limits<double>::epsilon()
+    ){
+      extraOffset = clock_data.TriggerTime()-clock_data.BeamGateTime();
+    }
+
+    double const triggerOffset = trigger_offset(clock_data) + clock_data.TPCClock().Ticks(extraOffset);
```

Comment:
Here is where the difference is.
Incidentally, it should be possible to check the existence of the trigger time with an equality comparison, since in case of no trigger the exact value from `numeric_limits` is assigned. But ok, more care won't hurt.